app/src/main/kotlin/com/wire/android/ui/connection/ConnectionActionButtonViewModel.kt

```diff
@@ -0,0 +1,212 @@
+import com.wire.android.appLogger
+import com.wire.android.navigation.BackStackMode
+import com.wire.android.navigation.EXTRA_CONNECTION_IGNORED_USER_NAME
+import com.wire.android.navigation.EXTRA_CONNECTION_STATE
+import com.wire.android.navigation.EXTRA_USER_ID
+import com.wire.android.navigation.EXTRA_USER_NAME
+import com.wire.android.navigation.NavigationCommand
+import com.wire.android.navigation.NavigationItem
+import com.wire.android.navigation.NavigationManager
+import com.wire.android.ui.common.snackbar.ShowSnackBarUseCase
+import com.wire.android.util.dispatchers.DispatcherProvider
+import com.wire.android.util.ui.UIText
+import com.wire.kalium.logic.data.id.QualifiedID
+import com.wire.kalium.logic.data.id.QualifiedIdMapper
+import com.wire.kalium.logic.data.id.toQualifiedID
+import com.wire.kalium.logic.data.user.ConnectionState
+import com.wire.kalium.logic.feature.connection.AcceptConnectionRequestUseCase
+import com.wire.kalium.logic.feature.connection.AcceptConnectionRequestUseCaseResult
+import com.wire.kalium.logic.feature.connection.CancelConnectionRequestUseCase
+import com.wire.kalium.logic.feature.connection.CancelConnectionRequestUseCaseResult
+import com.wire.kalium.logic.feature.connection.IgnoreConnectionRequestUseCase
+import com.wire.kalium.logic.feature.connection.IgnoreConnectionRequestUseCaseResult
+import com.wire.kalium.logic.feature.connection.SendConnectionRequestResult
+import com.wire.kalium.logic.feature.connection.SendConnectionRequestUseCase
+import com.wire.kalium.logic.feature.connection.UnblockUserResult
+import com.wire.kalium.logic.feature.connection.UnblockUserUseCase
+import com.wire.kalium.logic.feature.conversation.CreateConversationResult
+import com.wire.kalium.logic.feature.conversation.GetOrCreateOneToOneConversationUseCase
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+
+interface ConnectionActionButtonBaseViewModel {
+
+    fun state(): ConnectionState
+    fun onSendConnectionRequest()
+    fun onCancelConnectionRequest()
+    fun onAcceptConnectionRequest()
+    fun onIgnoreConnectionRequest()
+    fun onUnblockUser()
+    fun onOpenConversation()
+}
+
+@Suppress("LongParameterList", "TooManyFunctions")
+@HiltViewModel
+class ConnectionActionButtonViewModel @Inject constructor(
+    private val navigationManager: NavigationManager,
+    private val dispatchers: DispatcherProvider,
+    private val sendConnectionRequest: SendConnectionRequestUseCase,
+    private val cancelConnectionRequest: CancelConnectionRequestUseCase,
+    private val acceptConnectionRequest: AcceptConnectionRequestUseCase,
+    private val ignoreConnectionRequest: IgnoreConnectionRequestUseCase,
+    private val unblockUser: UnblockUserUseCase,
+    private val getOrCreateOneToOneConversation: GetOrCreateOneToOneConversationUseCase,
+    private val showSnackBarUseCase: ShowSnackBarUseCase,
```

Comment:
I find it weird to have a `UseCase` called by `ViewModel` in order to display UI stuff.

Why would the `ViewModel` trigger a `SnackBar`? It should be the UI responsibility to handle UI, not the `ViewModel`.

Can't we have this handled by the Composables and use something like `CompositionLocal` to pass SnackBar stuff when needed? _i.e._ when a Composable wants to display a `SnackBar`, it gets a `SnackBarManager` in its local composition tree in order to be displayed.

ViewModels should not worry about **HOW** things are displayed. They should worry about **WHAT** is displayed.

This pattern here is cementing the design to the ViewModel, something we do in some places (looking at Dialogs, specifically) but we should have avoided it since the beginning of the project.

If the ViewModel exposes a `Operation Failed` state. The Composable should figure out if that goes in a `Snackbar`, `Dialog`, `Toast`, or whatever are the design requirements.

Reply:
I think the only problem that may arrise if design decides to display "the message" ( because snackbar are only shown when we want to display a message, from my knowledge ) in that case we could create another use case for toast and another one for dialog, that would imply that we also handle dialogs from the viewmodel ( at least showing them ) if this would be working fine and create no problems I do not think it creates some danger for us ? However, if we would not be able to create a use case allowing us to simply swap the use case inside the viewmodel, than I think it could be cubersome for the future changes, because all the sudden we need to change "flow of data" between the viewmodel to the Composable element to handle the dialog or a toast there. 

What else could change from the design would be the "action"to be performed in place of displaying a message, but that could be either navigation, which we are already doing inside the viewmodel, or changing some data or showing some additional stuff on the UI, but that could be a simple flag exposed inside the state in case of additiional stuff displayed or a matter of consuming data, which we are already doing both.

Perhaps it would be troublesome if we want to consume the use case in some kind of services or a worker, but how many times do we do that ? And if we do, we could adjust the viewmodel by then. 

Am I missing something here ?

Reply:
Also if we let the component handle displaying of a dialog or a snackbar, it w ill not be a "stand-alone"component anymore it would have to have some dependency on a Scaffold snackbar state or either expose a lambda that would make it coupled to something from the outside world. 

Another thing that came to my mind is that the dialogs can also display snackbar/toast and another dialog. 

I personally thing enclosing everything that makes the component not coupled to anything else is a better approach IF we are able to have use case for displaying dialogs/toast/snackbars and really think it through if we make that decision that we are not shooting us in foot and have the solutions for all the cases we can think of, otherwise this may be riscy, because we will do a refactor of a refactor of a refactor ;P